Framework/Core/include/Framework/Array2D.h

```diff
@@ -114,6 +118,146 @@ struct Array2D {
+      rowmap{},
+      colmap{}
+  {
+  }
+
+  LabeledArray(T const* data, uint32_t rows, uint32_t cols, std::vector<std::string> labels_rows_ = {}, std::vector<std::string> labels_cols_ = {})
+    : values{data, rows, cols},
+      labels_rows{labels_rows_},
+      labels_cols{labels_cols_},
+      rowmap{},
+      colmap{}
+  {
+    if (labels_rows.empty() == false) {
+      assert(labels_rows.size() == rows);
+      for (auto i = 0u; i < labels_rows.size(); ++i) {
+        rowmap.emplace(labels_rows[i], (uint32_t)i);
+      }
+    }
+    if (labels_cols.empty() == false) {
+      assert(labels_cols.size() == cols);
+      for (auto i = 0u; i < labels_cols.size(); ++i) {
+        colmap.emplace(labels_cols[i], (uint32_t)i);
+      }
+    }
+  }
+
+  LabeledArray(T const* data, uint32_t size, std::vector<std::string> labels_ = {})
+    : values{data, 1, size},
+      labels_rows{},
+      labels_cols{labels_},
+      rowmap{},
+      colmap{}
+  {
+    if (labels_cols.empty() == false) {
+      assert(labels_cols.size() == size);
+      for (auto i = 0u; i < labels_cols.size(); ++i) {
+        colmap.emplace(labels_cols[i], (uint32_t)i);
+      }
+    }
+  }
+
+  LabeledArray(Array2D<T> const& data, std::vector<std::string> labels_rows_ = {}, std::vector<std::string> labels_cols_ = {})
+    : values{data},
+      labels_rows{labels_rows_},
+      labels_cols{labels_cols_},
+      rowmap{},
+      colmap{}
+  {
+    if (labels_rows.empty() == false) {
+      assert(labels_rows.size() == values.rows);
+      for (auto i = 0u; i < labels_rows.size(); ++i) {
+        rowmap.emplace(labels_rows[i], (uint32_t)i);
+      }
+    }
+    if (labels_cols.empty() == false) {
+      assert(labels_cols.size() == values.cols);
+      for (auto i = 0u; i < labels_cols.size(); ++i) {
+        colmap.emplace(labels_cols[i], (uint32_t)i);
+      }
+    }
+  }
+
+  LabeledArray(LabeledArray<T> const& other) = default;
+  LabeledArray(LabeledArray<T>&& other) = default;
+  LabeledArray& operator=(LabeledArray<T> const& other) = default;
+  LabeledArray& operator=(LabeledArray<T>&& other) = default;
+
+  ~LabeledArray() = default;
+
+  T get(uint32_t y, uint32_t x) const
+  {
+    return values(y, x);
+  }
+
+  T get(std::string y, std::string x) const
+  {
+    return values(rowmap.find(y)->second, colmap.find(x)->second);
+  }
+
+  T get(uint32_t x) const
+  {
+    return values[0][x];
+  }
+
+  T getRow(u_int32_t y) const
+  {
+    return values[y];
+  }
+
+  T* operator[](uint32_t y) const
+  {
+    return values[y];
+  }
+
+  auto getLabelsRows()
+  {
+    return labels_rows;
+  }
+
+  auto getLabelsCols()
+  {
+    return labels_cols;
+  }
+
+  auto getData()
+  {
+    return values;
+  }
+
+  auto rows()
+  {
+    return values.rows;
+  }
+
+  auto cols()
+  {
+    return values.cols;
+  }
+
+ private:
+  Array2D<T> values;
+  std::vector<std::string> labels_rows;
+  std::vector<std::string> labels_cols;
+  std::unordered_map<std::string, uint32_t> rowmap;
```

Comment:
do you really need the unordered_map?

Reply:
It was the easiest way to make label getter work without too much performance drop. I will try to rework this later.

Reply:
Did you check what was the actual performance issue? I strongly believe that in `get(std::string, std::string)` the most time consuming part is allocating the strings, not the lookup. Why not using `get(char const*, char const*)`?

Reply:
Getter arguments should've been references... The issue was that using `std::find` on string vector was slower than map lookup even for sizes below 10. And since typical usage would be getting non-sequential values of the array to use in multiple if statements in `process()` that could be a problem. But I do not like using `unordered_map` and intend to rewrite it with the hash trick we use in registries.

Reply:
even if they are references, if the common usage is get("foo", "bar"), as I assume it is, it will do two allocations for each invocation (because `char const*` is not a `string`). At least you should provide an overload, if that is a typical usecase.